File: gtci/01-two-pointers/cpp/7-dutch-national-flag-problem.cpp
#include <cassert>
#include <iostream>
#include <utility>
#include <vector>

using namespace std;

class Solution {
   public:
    vector<int> sort(vector<int> &&arr) {
        std::vector<int>::iterator lowItr = arr.begin();
        std::vector<int>::iterator highItr = arr.end() - 1;

        std::vector<int>::iterator itr = arr.begin();
        while (itr <= highItr) {
            if (*itr == 0) {
                std::swap(*itr, *lowItr);
                lowItr++;
                itr++;
            } else if (*itr == 1) {
                itr++;
            } else {
                std::swap(*itr, *highItr);
                highItr--;
            }
        }

        return arr;
    }
};

int main() {
    Solution sol;

    assert(sol.sort(vector<int>{0, 1, 2, 0, 1, 2}) ==
           vector<int>({0, 0, 1, 1, 2, 2}));
    assert(sol.sort(vector<int>{2, 2, 0, 1, 2, 0}) ==
           vector<int>({0, 0, 1, 2, 2, 2}));
    cout << "All test cases passed." << endl;

    return 0;
}

File: gtci/01-two-pointers/python3/7-dutch-national-flag-problem.py
class Solution:
    def sort(self, arr):
        low = 0  # where the 0s will go
        high = len(arr) - 1  # where the 2s will go

        idx = 0
        while idx <= high:
            if arr[idx] == 0:
                arr[idx], arr[low] = arr[low], arr[idx]
                low += 1
                idx += 1
            elif arr[idx] == 1:  # 1s stay in the middle
                idx += 1
            else:
                arr[idx], arr[high] = arr[high], arr[idx]
                high -= 1

        return arr


if __name__ == "__main__":
    sol = Solution()

    assert sol.sort([0, 1, 2, 0, 1, 2]) == [0, 0, 1, 1, 2, 2]
    assert sol.sort([2, 2, 0, 1, 2, 0]) == [0, 0, 1, 2, 2, 2]
    print("All test cases passed.")
